Data race in TestDirectoryListing

From https://github.com/tailscale/tailscale/actions/runs/8055884780/job/22003735121:

```
=== RUN   TestDirectoryListing
    tailfs_test.go:133: FileSystemForLocal listening at 127.0.0.1:36419
    tailfs_test.go:162: Remote for remote$%1 listening at 127.0.0.1:38807
    tailfs_test.go:169: FileServer for remote$%1 listening at 127.0.0.1:43629
==================
WARNING: DATA RACE
Read at 0x00c000046430 by goroutine 18:
  strings.(*Reader).Read()
      strings/reader.go:40 +0x55
  io.(*nopCloserWriterTo).Read()
      <autogenerated>:1 +0x6c
  net/http.(*readTrackingBody).Read()
      net/http/transport.go:662 +0x76
  io.copyBuffer()
      io/io.go:429 +0x29a
  io.(*multiReader).writeToWithBuffer()
      io/multi.go:54 +0x153
  io.(*multiReader).WriteTo()
      io/multi.go:45 +0x59
  io.copyBuffer()
      io/io.go:411 +0xd3
  io.CopyBuffer()
      io/io.go:402 +0x8f
  net/http.(*transferWriter).doBodyCopy()
      net/http/transfer.go:416 +0x144
  net/http.(*transferWriter).writeBody()
      net/http/transfer.go:360 +0xee9
  net/http.(*Request).write()
      net/http/request.go:755 +0x1413
  net/http.(*persistConn).writeLoop()
      net/http/transport.go:2451 +0x379
  net/http.(*Transport).dialConn.gowrap3()
      net/http/transport.go:[180](https://github.com/tailscale/tailscale/actions/runs/8055884780/job/22003735121#step:8:181)4 +0x33

Previous write at 0x00c000046430 by goroutine 9:
  strings.(*Reader).Seek()
      strings/reader.go:132 +0x16f
  github.com/studio-b12/gowebdav.(*authShim).Authorize.func1()
      github.com/studio-b12/gowebdav@v0.9.0/auth.go:241 +0x82
  github.com/studio-b12/gowebdav.(*Client).req()
      github.com/studio-b12/gowebdav@v0.9.0/requests.go:51 +0x575
  github.com/studio-b12/gowebdav.(*Client).propfind()
      github.com/studio-b12/gowebdav@v0.9.0/requests.go:84 +0x1a4
  github.com/studio-b12/gowebdav.(*Client).ReadDir()
      github.com/studio-b12/gowebdav@v0.9.0/client.go:161 +0x22f
  tailscale.com/tailfs/tailfsimpl.(*system).checkDirList()
      tailscale.com/tailfs/tailfsimpl/tailfs_test.go:260 +0x9d
  tailscale.com/tailfs/tailfsimpl.TestDirectoryListing()
      tailscale.com/tailfs/tailfsimpl/tailfs_test.go:48 +0xc7
  testing.tRunner()
      testing/testing.go:1689 +0x21e
  testing.(*T).Run.gowrap1()
      testing/testing.go:1742 +0x44

Goroutine 18 (running) created at:
  net/http.(*Transport).dialConn()
      net/http/transport.go:1804 +0x27fe
  net/http.(*Transport).dialConnFor()
      net/http/transport.go:1489 +0x124
  net/http.(*Transport).queueForDial.gowrap1()
      net/http/transport.go:1454 +0x44

Goroutine 9 (running) created at:
  testing.(*T).Run()
      testing/testing.go:1742 +0x825
  testing.runTests.func1()
      testing/testing.go:2161 +0x85
  testing.tRunner()
      testing/testing.go:1689 +0x21e
  testing.runTests()
      testing/testing.go:2159 +0x8be
  testing.(*M).Run()
      testing/testing.go:2027 +0xf17
  main.main()
      _testmain.go:51 +0x2bd
==================
    tailfs_test.go:162: Remote for _remote$%2 listening at 127.0.0.1:42385
    tailfs_test.go:169: FileServer for _remote$%2 listening at 127.0.0.1:36391
[202](https://github.com/tailscale/tailscale/actions/runs/8055884780/job/22003735121#step:8:203)4/02/26 21:34:46 http: proxy error: net/http: timeout awaiting response headers
2024/02/26 21:34:46 serve: invalid argument
    tailfs_test.go:138: Accept: accept tcp 127.0.0.1:36419: use of closed network connection
    testing.go:1398: race detected during execution of test
--- FAIL: TestDirectoryListing (5.09s)
FAIL	tailscale.com/tailfs/tailfsimpl
```

cc @oxtoacart

--- a/tailfs/tailfsimpl/tailfs_test.go
+++ b/tailfs/tailfsimpl/tailfs_test.go
@@ -5,6 +5,7 @@ package tailfsimpl
 
 import (
 	"fmt"
+	"io"
 	"io/fs"
 	"log"
 	"net"
@@ -142,7 +143,7 @@ func newSystem(t *testing.T) *system {
 		}
 	}()
 
-	client := gowebdav.NewClient(fmt.Sprintf("http://%s", l.Addr()), "", "")
+	client := gowebdav.NewAuthClient(fmt.Sprintf("http://%s", l.Addr()), &noopAuthorizer{})
 	client.SetTransport(&http.Transport{DisableKeepAlives: true})
 	s := &system{
 		t:       t,
@@ -374,4 +375,34 @@ func fileInfoToStatic(fi fs.FileInfo, fixupMode bool) fs.FileInfo {
 
 func pathTo(remote, share, name string) string {
 	return path.Join(domain, remote, share, name)
+}
+
+// noopAuthorizer implements gowebdav.Authorizer. It does no actual
+// authorizing. We use it in place of gowebdav's built-in authorizer in order
+// to avoid a race condition in that authorizer.
+type noopAuthorizer struct{}
+
+func (a *noopAuthorizer) NewAuthenticator(body io.Reader) (gowebdav.Authenticator, io.Reader) {
+	return &noopAuthenticator{}, nil
+}
+
+func (a *noopAuthorizer) AddAuthenticator(key string, fn gowebdav.AuthFactory) {
+}
+
+type noopAuthenticator struct{}
+
+func (a *noopAuthenticator) Authorize(c *http.Client, rq *http.Request, path string) error {
+	return nil
+}
+
+func (a *noopAuthenticator) Verify(c *http.Client, rs *http.Response, path string) (redo bool, err error) {
+	return false, nil
+}
+
+func (a *noopAuthenticator) Clone() gowebdav.Authenticator {
+	return &noopAuthenticator{}
+}
+
+func (a *noopAuthenticator) Close() error {
+	return nil
 }

